File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        self.result = False

        queue = deque([root])

        while queue:
            node = queue.popleft()

            if self.is_subtree(node, subRoot):
                return True

            if node.left:
                queue.append(node.left)
            
            if node.right:
                queue.append(node.right)
        
        return False
    
    def is_subtree(self, root, sub_root):
        if not root and not sub_root:
            return True
        if root and sub_root and root.val == sub_root.val:
            return self.is_subtree(root.left, sub_root.left) and self.is_subtree(root.right, sub_root.right)
        else:
            return False

File: Data Structures & Algorithms/same-binary-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool result = true;
    bool isSameTree(TreeNode* p, TreeNode* q) {
        dfs(p, q);

        return result;
    }

    void dfs(TreeNode* t1, TreeNode* t2) {
        if (!t1 && !t2) {
            return;
        }

        if ((!t1 && t2) || (!t2 && t1) || (t1->val != t2->val)) {
            result = false;
            return;
        }

        dfs(t1->left, t2->left);
        dfs(t1->right, t2->right);
    }
};
